File: ios/Outbound/Outbound/Goals/GoalProgressEngine.swift
import Foundation

enum GoalProgressEngine {
    static func makeProgress(
        goal: GoalDefinition,
        activities: [SavedActivity],
        addingSummary summary: ActivitySummary? = nil,
        calendar: Calendar = .current
    ) -> GoalProgressSnapshot {
        let weekEnd = calendar.date(byAdding: .day, value: 7, to: goal.weekStart) ?? goal.weekStart
        let inRangeActivities = activities.filter { activity in
            activity.startedAt >= goal.weekStart && activity.startedAt < weekEnd
        }

        let baseValue: Int
        switch goal.kind {
        case .weeklySessions:
            baseValue = inRangeActivities.count
        case .weeklyMinutes:
            let totalSeconds = inRangeActivities.reduce(0) { $0 + $1.durationSecs }
            baseValue = Int(ceil(Double(totalSeconds) / 60.0))
        }

        let extraValue: Int
        switch goal.kind {
        case .weeklySessions:
            extraValue = summary == nil ? 0 : 1
        case .weeklyMinutes:
            if let summary {
                extraValue = Int(ceil(Double(summary.durationSecs) / 60.0))
            } else {
                extraValue = 0
            }
        }

        let currentValue = baseValue + extraValue
        let remainingValue = max(0, goal.targetValue - currentValue)
        let isComplete = currentValue >= goal.targetValue
        let percentComplete = min(1, goal.targetValue > 0 ? Double(currentValue) / Double(goal.targetValue) : 1)

        return GoalProgressSnapshot(
            goal: goal,
            currentValue: currentValue,
            targetValue: goal.targetValue,
            percentComplete: percentComplete,
            isComplete: isComplete,
            remainingValue: remainingValue,
            summaryLine: summaryLine(kind: goal.kind, currentValue: currentValue, targetValue: goal.targetValue),
            coachLine: coachLine(kind: goal.kind, remainingValue: remainingValue, isComplete: isComplete)
        )
    }

    static func suggestedTarget(
        theme: GoalFocusTheme,
        activities: [SavedActivity],
        phase: MotivationPhase,
        calendar: Calendar = .current,
        now: Date = Date()
    ) -> Int {
        let recentWeekCount = activitiesThisWeek(activities: activities, calendar: calendar, now: now)

        switch theme {
        case .lightMovement:
            if recentWeekCount >= 3 || phase == .momentum { return 45 }
            return 20
        case .comeback:
            return recentWeekCount >= 2 ? 3 : 2
        case .consistency:
            if phase == .momentum { return 4 }
            if recentWeekCount >= 2 { return 3 }
            return 2
        }
    }

    private static func activitiesThisWeek(
        activities: [SavedActivity],
        calendar: Calendar,
        now: Date
    ) -> Int {
        guard let week = calendar.dateInterval(of: .weekOfYear, for: now) else { return 0 }
        return activities.filter { week.contains($0.startedAt) }.count
    }

    private static func summaryLine(kind: GoalKind, currentValue: Int, targetValue: Int) -> String {
        switch kind {
        case .weeklySessions:
            return "\(currentValue) of \(targetValue) sessions this week"
        case .weeklyMinutes:
            return "\(currentValue) of \(targetValue) min this week"
        }
    }

    private static func coachLine(kind: GoalKind, remainingValue: Int, isComplete: Bool) -> String {
        if isComplete {
            return "You completed this week's focus."
        }

        switch kind {
        case .weeklySessions:
            if remainingValue == 1 {
                return "One more session would finish this week's focus."
            }
            return "\(remainingValue) more sessions would finish this week's focus."
        case .weeklyMinutes:
            return "\(remainingValue) more minutes would finish this week's focus."
        }
    }
}

File: android/core/model/src/main/kotlin/com/plainstride/outbound/core/model/activity/GoalProgressEngine.kt
package com.plainstride.outbound.core.model.activity

import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import kotlin.math.ceil

enum class GoalKind { weeklySessions, weeklyMinutes }
enum class GoalFocusTheme { consistency, comeback, lightMovement }
enum class MotivationPhase { firstSession, steady, comeback, momentum, completedToday }

data class GoalDefinition(
    val kind: GoalKind,
    val theme: GoalFocusTheme,
    val targetValue: Int,
    val weekStart: Instant,
)

data class GoalProgressSnapshot(
    val currentValue: Int,
    val targetValue: Int,
    val percentComplete: Double,
    val isComplete: Boolean,
    val remainingValue: Int,
)

/** Progress semantics only. Localized summary and guide copy belongs to the presentation layer. */
object GoalProgressEngine {
    fun makeProgress(
        goal: GoalDefinition,
        activities: List<ActivityFacts>,
        addingSummary: ActivitySummaryFacts? = null,
        zoneId: ZoneId,
    ): GoalProgressSnapshot {
        val weekEnd = ZonedDateTime.ofInstant(goal.weekStart, zoneId).plusDays(7).toInstant()
        val inRange = activities.filter { !it.startedAt.isBefore(goal.weekStart) && it.startedAt.isBefore(weekEnd) }
        val baseValue = when (goal.kind) {
            GoalKind.weeklySessions -> inRange.size
            GoalKind.weeklyMinutes -> ceil(inRange.sumOf { it.durationSeconds } / 60.0).toInt()
        }
        val extraValue = when (goal.kind) {
            GoalKind.weeklySessions -> if (addingSummary == null) 0 else 1
            GoalKind.weeklyMinutes -> addingSummary?.let { ceil(it.durationSeconds / 60.0).toInt() } ?: 0
        }
        val currentValue = baseValue + extraValue
        return GoalProgressSnapshot(
            currentValue = currentValue,
            targetValue = goal.targetValue,
            percentComplete = minOf(1.0, if (goal.targetValue > 0) currentValue.toDouble() / goal.targetValue else 1.0),
            isComplete = currentValue >= goal.targetValue,
            remainingValue = maxOf(0, goal.targetValue - currentValue),
        )
    }

    fun suggestedTarget(
        theme: GoalFocusTheme,
        activities: List<ActivityFacts>,
        phase: MotivationPhase,
        weekStart: Instant,
        weekEnd: Instant,
    ): Int {
        val recentWeekCount = activities.count { !it.startedAt.isBefore(weekStart) && it.startedAt.isBefore(weekEnd) }
        return when (theme) {
            GoalFocusTheme.lightMovement -> if (recentWeekCount >= 3 || phase == MotivationPhase.momentum) 45 else 20
            GoalFocusTheme.comeback -> if (recentWeekCount >= 2) 3 else 2
            GoalFocusTheme.consistency -> when { phase == MotivationPhase.momentum -> 4; recentWeekCount >= 2 -> 3; else -> 2 }
        }
    }
}
